mlton/backend/ssa2-to-rssa.fun

```diff
@@ -1642,38 +1643,255 @@ fun convert (program as S.Program.T {functions, globals, main, ...},
                           returns = transTypes returns,
                           start = start}
          end
+      fun translateGlobalStatics statements
+         : (Var.t * Type.t * Var.t Static.t) vector * S.Statement.t vector
+         =
+         let
+            val statics = ref []
+            val keeps = ref []
+            datatype elem = datatype Static.Data.elem
+            datatype z = datatype PackedRepresentation.staticOrElem
+            val {get = globalStatic: Var.t -> Var.t elem option,
+                 set = setGlobalStatic, destroy = destGlobalStatics} =
+               Property.destGetSetOnce
+               (Var.plist, Property.initConst NONE)
+            fun pushStatic (v, ty, rty, s) =
+               (setVarInfo (v, {ty=ty});
+                List.push (statics, (v, rty, s)))
+            fun pushKeep st = List.push (keeps, st)
+
+            fun validArg {elt, isMutable} =
+               case !Control.staticAllocInternalPtrs of
+                    Control.All => true
+                  | Control.Static =>
+                       not (isMutable andalso Option.exists (toRtype elt, Type.isObjptr))
+                  | Control.None => false
+            fun validCon con =
+              case (con,
+                    !Control.staticAllocObjects,
+                    !Control.staticAllocVectors) of
+                   (S.ObjectCon.Con _, true, _) => true
+                 | (S.ObjectCon.Tuple, true, _) => true
+                 | (S.ObjectCon.Sequence, _, true) => true
+                 | _ => false
+            fun getLocation (ty, isEmpty) =
+               case S.Type.dest ty of
+                    (* Important: isMutable is a proxy for hasIdentity,
+                     * so we need to preserve it even if there are no fields *)
+                    S.Type.Object {args, con} =>
+                     if validCon con andalso
+                        (isEmpty orelse
+                           Vector.forall
+                           (S.Prod.dest args, validArg))
+                     then
+                        if S.Prod.someIsMutable args
+                           then Static.MutStatic
+                        else Static.ImmStatic
+                     else
+                        Static.Heap
```

Comment:
As before, this is stating whether an object can be statically _allocated_, as in go to a static location. The validity check is for checking whether it is sound to have it outside the heap on our GC. Regardless, static initialization may be performed irrespective of this choice unless there's a reason to avoid it.

Reply:
O.k.  I was confused by the relationship with https://github.com/MLton/mlton/pull/328#discussion_r321527949, because it seemed that here `Static.Heap` is being returned just to signal that the object should not be statically initialized.  But, you are arguing that those are orthogonal choices.

Reply:
Yeah first we make the decision that it must not be statically allocated, then we make the optimization that it should not be statically initialized.